### PR TITLE
Refactor budgeting to make the logic cleaner

### PR DESCRIPTION
**Pull request by kdnilsen**
This commit consolidates computations related to budgeting of reserves for old-gen evacuation and old-gen promotion within shenandoahGeneration::prepare_regions_and_collection_set().  Previously, these computations had been scattered between several functional units, including shenandoahHeuristics::choose_collection_set() and shenandoahOldHeuristics::prime_collection_set().  A previous pull request brought some of the redundant functionality into prepare_regions_and_collection_set().  This pull request removes the redundant computations from the other locations.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Change must be properly reviewed (1 review required, with at least 1 [Committer](https://openjdk.org/bylaws#committer))

### Reviewers
 * [Roman Kennke](https://openjdk.org/census#rkennke) (@rkennke - **Reviewer**) ⚠️ Review applies to [be1fb125](https://git.openjdk.org/shenandoah/pull/148/files/be1fb125ea0c2856f38f165d71f64baa2b7d573b)
 * [William Kemper](https://openjdk.org/census#wkemper) (@earthling-amzn - Committer)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/shenandoah pull/148/head:pull/148` \
`$ git checkout pull/148`

Update a local copy of the PR: \
`$ git checkout pull/148` \
`$ git pull https://git.openjdk.org/shenandoah pull/148/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 148`

View PR using the GUI difftool: \
`$ git pr show -t 148`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/shenandoah/pull/148.diff">https://git.openjdk.org/shenandoah/pull/148.diff</a>

</details>

### CODE REVIEW

File: src/hotspot/share/gc/shenandoah/shenandoahHeap.inline.hpp [hunk 1]
```diff
@@ -686,6 +686,7 @@ size_t ShenandoahHeap::capture_old_usage(size_t old_usage) {
 }
 
 void ShenandoahHeap::set_previous_promotion(size_t promoted_bytes) {
+  shenandoah_assert_heaplocked();
```
Comment (rkennke) [inline src/hotspot/share/gc/shenandoah/shenandoahHeap.inline.hpp:689]: Is this related and/or relevant?

Comment (kdnilsen) [reply src/hotspot/share/gc/shenandoah/shenandoahHeap.inline.hpp:689]: We can remove that.  It was "somewhat" related to the historical progression that eventually resulted in this refactoring code.  At one point in my development, I had accidentally tried to set_previous_promotion() without having the heap lock, and the results were unpredictable.

Comment (rkennke) [reply src/hotspot/share/gc/shenandoah/shenandoahHeap.inline.hpp:689]: Hmm ok. No, leave it there. If it checks an important condition, then it should be useful.

**Review by rkennke — APPROVED** [review]

File: src/hotspot/share/gc/shenandoah/shenandoahGeneration.cpp [hunk 2]
```diff
@@ -221,7 +221,352 @@ void ShenandoahGeneration::prepare_gc(bool do_old_gc_bootstrap) {
   }
 }
 
-void  ShenandoahGeneration::prepare_regions_and_collection_set(bool concurrent) {
+void ShenandoahGeneration::compute_evacuation_budgets(ShenandoahHeap* heap, ShenandoahCollectionSet* collection_set,
```
Comment (ysramakrishna) [inline src/hotspot/share/gc/shenandoah/shenandoahGeneration.cpp:224]: The refactoring of the large chunks of code into the smaller methods is definitely a step in the right direction in making the code easier to follow.

Nit: Would it make the code more readable to consolidate the long sections of comments into a block comment at the top of each of the methods and have much shorter comments in the body of the methods?

File: src/hotspot/share/gc/shenandoah/shenandoahGeneration.hpp [hunk 3]
```diff
@@ -111,6 +111,16 @@ class ShenandoahGeneration : public CHeapObj<mtGC> {
   // Used by concurrent and degenerated GC to reset regions.
   void prepare_gc(bool do_old_gc_bootstrap);
 
+  // Compute evacuation budgets prior to choosing collection set.
```
Comment (ysramakrishna) [inline src/hotspot/share/gc/shenandoah/shenandoahGeneration.hpp:114]: Do these methods need to be public? They looked more like helper methods in the way they are currently being used.

Another thing to think about is whether the reference parameters in one signature and regular ones in the other are potential sources of confusion or error in the future. I can see why they are like they are in their current use, but might indicate that perhaps a cleaner refactoring of the methods and the state that it uses or modifies, and where that state should reside.

Nothing to do at the moment, but worth thinking about.

**Review by ysramakrishna — COMMENTED** [review]
Left a few minor syntactic/stylistic comments.

File: src/hotspot/share/gc/shenandoah/shenandoahGeneration.cpp (hunk 2 above)
Comment (kdnilsen) [reply src/hotspot/share/gc/shenandoah/shenandoahGeneration.cpp]: Thanks for your suggestion.  Maybe you can do a subsequent pass (in a future PR) to clean up the structure of comments.  I'm sure these can be made better.  Sometimes, a "reader" of the comments has the best perspective on how to make the comments more valuable to other readers.

File: src/hotspot/share/gc/shenandoah/shenandoahGeneration.hpp (hunk 3 above)
Comment (kdnilsen) [reply src/hotspot/share/gc/shenandoah/shenandoahGeneration.hpp:114]: Agree with suggestion to make compute_evacuation_budgets and adjust_evacuation_budgets private.  Done.

I welcome further refactoring improvements in the future.  Am eager to get this code integrated now (as is) so that we can proceed with other performance improvements.

**Review by earthling-amzn — APPROVED** [review]
